compartmentalize-grains-via-pop.md

```diff
@@ -0,0 +1,71 @@
+maintaining backwards compatibility.
+
+# Motivation
+[motivation]: #motivation
+
+"The original implementation of grains was not designed to scale to the level that salt requires" - thatch45
+Tom wrote POP to solve this problem and we're going to use it. 
+
+By using Plugin Oriented Programming, grains can be separated into individual projects that can be maintained
+by the community safely.
+
+This also solves the issue of `core.py` being an unmaintainable monolith.  It allows platform specific grains
+to be independently maintained.
+
+Working group leaders and other community members focused on specific platforms can manage the release cycle for
+individual grains projects independently from salt.
+
+# Design
+[design]: #detailed-design
+
+Grains are currently created by calling functions in the salt loader.  We would simply need to call
+external POP functions to collect grains.  We need to maintain backwards compatibility with custom user grains.
+
+- All salt grains have already been ported to and tested on the new platform.
+- The new platform has many more features than the salt-grains implementation, but is compatible with salt grains.
+- None of the grain names have been changed, but the output of grains is more consistent/complete between platforms.
+- This will resolve many bugs that are open against grains and will migrate those bugs OUT of the salt
+platform -- making salt easier to develop
+- Grains in the new platform are independently and fully tested.
+- Most grains tests will be able to be removed from salt; speeding up it's test suite.
+- This comprises a complete rewrite of ALL grains.  Backwards compatibility is expected, but we need this to get
+in early so that any unintentional issues can be resolved quickly.
+- Salt releases will be able to pin idem-grain dependencies with each release.
+- Users will be able to update their grains implementation to a custom idem-grain version.
+
+## Alternatives
+[alternatives]: #alternatives
+
+Keep things the way they are.  Does this mean we would suffer in stagnation?  Will we be stuck maintaining a monolith.
+
+## Unresolved questions
+[unresolved]: #unresolved-questions
+
+What will further adoption of POP mean inside of salt?  The purpose of this sep and these features is not to provide
+a definitive answer to this question -- rather to introduce plugin oriented programming into salt in a very safe way.
+In the future this will determine the design so that salt can be further compartmentalized and written in POP.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+Why should we *not* do this? Please consider:
+
+- We will be introducing a number of new dependencies.
+- The new implementation of grains/pop is written with python3.6 and later with asyncio
+(The latest release of salt is also python3.6 and later).
+- Some grains for specific OSes (such as SUSE) implement some grains differently from other platforms.  We need
```

Comment:
This sounds like the biggest problem. Even though you say splitting them up has made them more consistent, it would be a lot easier to maintain if they were all in the same place.

How is this going to work for grains whose implementations differ not based on OS? For example based on whether particular services, package managers, or virtualisations are being used? There's a big matrix of how the same grain should get its data on different minions.

Reply:
damon's solution below is perfect for this.  Something like an `idem-posix` that contains all  shared code will keep maintainability simple.

Reply:
I don't see how? This isn't about duplicating shared code, it's about making sure different code stays consistent.

And then how do you find which project is implementing your grain? Is it in `idem-posix` or overridden by `idem-linux` or by `idem-debian` or `idem-ubuntu` or `idem-ec2` or `idem-networkmanager` etc?

It's far easier to get everything right if everything for a single grain is in a single location.

Reply:
One of the things that we've also been seeing -- at least I have -- is something that we had in the past but definitely haven't been keeping up, and that's maturity.

Do we have any kind of formalized patterns around maturity for corn grains? I think this will become *essential* as we explore pop-style dependencies. Some users will be comfortable with rapidly changing and experimental grains/features, others will definitely only want to use mature/stable features that aren't rapidly changing.

Reply:
Grains in grainsv2 aren't easily overwritten.  We thought of this early on and each grain can only be set once.  If `idem-debian` tries to overwrite a grain an error will be thrown.  It's certainly  *possible*, but not easy,  to overwrite a grain.

As for maturity, the new grains will be gated and disabled by default.  Users that enjoy the stability of grains as they know and love them will still be able to have that.  Grainsv2 will be able to gain maturity through those who choose to enable them.

I will also note that grains on platforms like BSD, solaris, and AIX, are fully tested in their idem-platform projects and have absolutely no running tests in salt.